File: csharp/ToolGood.Algorithm/Internals/Functions/Function_3.cs
using System;
using System.Text;

namespace ToolGood.Algorithm.Internals.Functions
{
	internal abstract class Function_3 : Function_2
	{
		protected FunctionBase func3;

		protected Function_3(FunctionBase[] funcs) : base(funcs)
		{
			if(funcs.Length >= 3) {
				this.func3 = funcs[2];
			}
		}

		protected override void AddFunction(StringBuilder stringBuilder, string functionName)
		{
			stringBuilder.Append(functionName);
			stringBuilder.Append('(');
			func1.ToString(stringBuilder, false);
			if(func2 != null) {
				stringBuilder.Append(", ");
				func2.ToString(stringBuilder, false);
				if(func3 != null) {
					stringBuilder.Append(", ");
					func3.ToString(stringBuilder, false);
				}
			}
			stringBuilder.Append(')');
		}


		#region Get_3
		protected Operand GetText_3(AlgorithmEngine engine, Func<AlgorithmEngine, string, Operand> tempParameter)
		{
			var args3 = func3.Evaluate(engine, tempParameter);
			if(args3.IsText) return args3;
			return ConvertToText(args3, 3);
		}

		protected Operand GetNumber_3(AlgorithmEngine engine, Func<AlgorithmEngine, string, Operand> tempParameter)
		{
			var args3 = func3.Evaluate(engine, tempParameter);
			if(args3.IsNumber) return args3;
			return ConvertToNumber(args3, 3);
		}

		protected Operand GetDate_3(AlgorithmEngine engine, Func<AlgorithmEngine, string, Operand> tempParameter)
		{
			var args3 = func3.Evaluate(engine, tempParameter);
			if(args3.IsDate) return args3;
			return ConvertToDate(args3, 3);
		}

		protected Operand GetBoolean_3(AlgorithmEngine engine, Func<AlgorithmEngine, string, Operand> tempParameter)
		{
			var args3 = func3.Evaluate(engine, tempParameter);
			if(args3.IsBoolean) return args3;
			return ConvertToBoolean(args3, 3);
		}
		protected Operand GetArray_3(AlgorithmEngine engine, Func<AlgorithmEngine, string, Operand> tempParameter)
		{
			var args3 = func3.Evaluate(engine, tempParameter);
			if(args3.IsArray) return args3;
			return ConvertToArray(args3, 3);
		}
		#endregion

	}


}

File: java/toolgood.algorithm/src/main/java/toolgood/algorithm/internals/Functions/Function_3.java
package toolgood.algorithm.internals.functions;

import java.lang.StringBuilder;

abstract class Function_3 extends FunctionBase {
    protected FunctionBase func1;
    protected FunctionBase func2;
    protected FunctionBase func3;

    protected Function_3(FunctionBase func1, FunctionBase func2, FunctionBase func3) {
        this.func1 = func1;
        this.func2 = func2;
        this.func3 = func3;
    }
    
    protected void addFunction(StringBuilder stringBuilder, String functionName) {
        stringBuilder.append(functionName);
        stringBuilder.append('(');
        func1.toString(stringBuilder, false);
        if (func2 != null) {
            stringBuilder.append(", ");
            func2.toString(stringBuilder, false);
            if (func3 != null) {
                stringBuilder.append(", ");
                func3.toString(stringBuilder, false);
            }
        }
        stringBuilder.append(')');
    }
}
